File: archived/src/c/ch01/1.3/l2/totient.c
#include "totient.h"

uint64_t phi_u64(uint64_t n) {
    if (n == 0) return 0; /* undefined, but keep 0 for safety */
    uint64_t result = n;
    uint64_t x = n;
    uint64_t p = 2;
    while (p * p <= x) {
        if (x % p == 0) {
            while (x % p == 0) x /= p;
            result -= result / p;
        }
        p += (p == 2 ? 1 : 2); /* after 2, check odds */
    }
    if (x > 1) {
        result -= result / x;
    }
    return result;
}

File: archived/src/python/ch01/1.3/l2/totient.py
def phi(n: int) -> int:
    """
    Euler’s totient φ(n): count of 1<=k<=n that are coprime to n.
    Teaching version: trial division factorization, O(sqrt n).
    """
    if n <= 0:
        raise ValueError("n must be positive")
    result = n
    x = n
    p = 2
    while p * p <= x:
        if x % p == 0:
            while x % p == 0:
                x //= p
            result -= result // p
        p += 1 if p == 2 else 2  # after 2, check odds only
    if x > 1:  # leftover prime
        result -= result // x
    return result